File: frontend/src/utils/ocrParser.ts
export interface ParsedPackage {
  name: string;
  dosage: string;
  instructions: string;
  expiryDate: string;
  supplyCount: number;
}

export const SAMPLE_PACKAGES: ParsedPackage[] = [
  {
    name: 'Aspirin Cardio',
    dosage: '81mg',
    instructions: 'Take 1 tablet daily with morning meal',
    expiryDate: '2027-12-31',
    supplyCount: 30,
  },
  {
    name: 'Lisinopril',
    dosage: '10mg',
    instructions: 'Take 1 tablet in the morning for blood pressure',
    expiryDate: '2026-11-30',
    supplyCount: 30,
  },
  {
    name: 'Metformin',
    dosage: '500mg',
    instructions: 'Take 1 tablet twice daily with food',
    expiryDate: '2028-06-30',
    supplyCount: 60,
  },
  {
    name: 'Atorvastatin',
    dosage: '20mg',
    instructions: 'Take 1 tablet at bedtime',
    expiryDate: '2025-01-15', // Past date intentionally to demonstrate expired safety alert
    supplyCount: 20,
  },
];

export function parseRawOcr(text: string): ParsedPackage {
  const lines = text
    .split('\n')
    .map((l) => l.trim())
    .filter((l) => l.length > 0);

  let name = lines.length > 0 ? lines[0].substring(0, 30) : 'Prescription Medicine';
  let dosage = '100mg';
  let instructions = 'Take as directed by doctor';
  let expiryDate = '2027-12-31';
  let supplyCount = 30;

  const dosageMatch = text.match(/(\d+\s*(?:mg|ml|mcg|g|tablets|capsules|pills))/i);
  if (dosageMatch) {
    dosage = dosageMatch[0];
  }

  const expiryMatch = text.match(/(?:exp[:\s]*|expires[:\s]*)([0-9]{1,2}[/-][0-9]{2,4}|[0-9]{4}[/-][0-9]{1,2})/i);
  if (expiryMatch) {
    expiryDate = expiryMatch[1];
  }

  const instructionLines = lines.filter((line) => {
    const upper = line.toUpperCase();
    return (
      upper.includes('TAKE') ||
      upper.includes('DAILY') ||
      upper.includes('AFTER') ||
      upper.includes('BEFORE') ||
      upper.includes('WITH') ||
      upper.includes('MORNING') ||
      upper.includes('NIGHT') ||
      upper.includes('MEAL')
    );
  });

  if (instructionLines.length > 0) {
    instructions = instructionLines.join('. ');
  }

  return {
    name,
    dosage,
    instructions,
    expiryDate,
    supplyCount,
  };
}

File: android-native/app/src/main/java/com/example/myapplication1/data/util/OcrParser.kt
package com.example.myapplication1.data.util

data class ParsedMedicineInfo(
    val name: String,
    val dosage: String,
    val instructions: String,
    val expiryDate: String,
    val supplyCount: Int = 30
)

object OcrParser {
    val samplePackages = listOf(
        ParsedMedicineInfo(
            name = "Aspirin",
            dosage = "500mg",
            instructions = "Take 1 tablet daily after food",
            expiryDate = "12/2027",
            supplyCount = 30
        ),
        ParsedMedicineInfo(
            name = "Lisinopril",
            dosage = "10mg",
            instructions = "Take 1 tablet in the morning",
            expiryDate = "06/2026",
            supplyCount = 30
        ),
        ParsedMedicineInfo(
            name = "Metformin",
            dosage = "850mg",
            instructions = "Take 1 tablet twice daily with meals",
            expiryDate = "03/2028",
            supplyCount = 60
        ),
        ParsedMedicineInfo(
            name = "Atorvastatin",
            dosage = "20mg",
            instructions = "Take 1 tablet at bedtime",
            expiryDate = "09/2026",
            supplyCount = 30
        )
    )

    fun parseOcrText(rawText: String): ParsedMedicineInfo {
        val lines = rawText.lines().map { it.trim() }.filter { it.isNotBlank() }
        
        var name = "Prescription Medicine"
        var dosage = "100mg"
        var instructions = "Take as directed by doctor"
        var expiryDate = "12/2026"
        var supplyCount = 30

        if (lines.isNotEmpty()) {
            name = lines.first().take(30)
        }

        val dosageRegex = Regex("""(\d+\s*(?:MG|ML|MCG|G|TABLETS|CAPSULES|PILLS|DROPS|PUFFS))""", RegexOption.IGNORE_CASE)
        dosageRegex.find(rawText)?.let {
            dosage = it.value.trim()
        }

        val expiryRegex = Regex("""(?:EXP[:\s]*|EXPIRES[:\s]*)([0-9]{1,2}[/-][0-9]{2,4}|[0-9]{4}[/-][0-9]{1,2})""", RegexOption.IGNORE_CASE)
        expiryRegex.find(rawText)?.let {
            expiryDate = it.groupValues.getOrNull(1) ?: it.value
        }

        val instructionLines = lines.filter { line ->
            val upper = line.uppercase()
            upper.contains("TAKE") || upper.contains("DAILY") || upper.contains("AFTER") || 
            upper.contains("BEFORE") || upper.contains("WITH") || upper.contains("MORNING") || 
            upper.contains("NIGHT") || upper.contains("MEAL") || upper.contains("HOURS")
        }
        if (instructionLines.isNotEmpty()) {
            instructions = instructionLines.joinToString(". ")
        }

        return ParsedMedicineInfo(
            name = name.replaceFirstChar { it.uppercase() },
            dosage = dosage,
            instructions = instructions,
            expiryDate = expiryDate,
            supplyCount = supplyCount
        )
    }
}
